openedx/core/djangoapps/import_export/courses/views.py

```diff
@@ -0,0 +1,528 @@
+                # generic error message
+                pass
+
+            return self._export_error_response(
+                {
+                    "context_course": str(courselike_module.location),
+                    "error": True,
+                    "error_message": str(exc),
+                    "failed_module":
+                    str(failed_item.location) if failed_item else None,
+                    "unit":
+                    str(unit.location) if unit else None
+                },
+                redirect_url=redirect_url
+            )
+        except Exception as exc:  # pylint: disable=broad-except
+            log.exception(
+                'There was an error exporting course %s',
+                courselike_key
+            )
+            return self._export_error_response(
+                {
+                    "context_course": courselike_module.url_name,
+                    "error": True,
+                    "error_message": str(exc),
+                    "unit": None
+                },
+                redirect_url=redirect_url
+            )
+
+        # The course is all set; return the tar.gz
+        wrapper = FileWrapper(export_file)
+
+        response = HttpResponse(wrapper, content_type='application/x-tgz')
+        response['Content-Disposition'] = 'attachment; filename={}'.format(
+            os.path.basename(
+                export_file.name.encode('utf-8')
+            )
+        )
+        response['Content-Length'] = os.path.getsize(export_file.name)
+        return response
+
+    @ensure_valid_course_key
+    @renderer_classes_decorator((JSONRenderer,))
+    def post(self, request, course_key_string):
+        """
+        The restful handler for importing a course.
+
+        GET
+            json: return json import status
+        POST or PUT
+            json: import a course via the .tar.gz file specified inrequest.FILES
+        """
+        courselike_key = CourseKey.from_string(course_key_string)
+        library = isinstance(courselike_key, LibraryLocator)
+
+        if library:
+            root_name = LIBRARY_ROOT
+            import_func = import_library_from_xml
+        else:
+            root_name = COURSE_ROOT
+            import_func = import_course_from_xml
+
+        filename = request.FILES['course-data'].name
+        courselike_string = unicode(courselike_key) + filename
+        data_root = path(settings.GITHUB_REPO_ROOT)
+        subdir = base64.urlsafe_b64encode(repr(courselike_key))
+        course_dir = data_root / subdir
+
+        status_key = "import_export.import.status:{}|{}".format(
+            request.user.username,
+            courselike_string
+        )
+
+        # Do everything in a try-except block to make sure everything is
+        # properly cleaned up.
+        try:
+            # Cache the import progress
+            self._save_request_status(request, courselike_string, 0)
```

Comment:
Since this is just shown back to a particular user, maybe it should just go into the user's session, rather than directly into the cache?